### PR TITLE
Sysinfo hotfixes registry enumeration memory peak

### LINKED ISSUE #9386: [Syscollector module] High memory peak when scanning Windows hotfixes
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.2-rc10 | Syscollector | Agent | Any | Windows |

Hi team,

Using Windows Performance Analyzer, we saw that retrieving hotfixes consumes 1MB avg peak. After a deeper analysis, we find that the enumeration of the subkeys of the registries where the installed hotfixes are searched returns a vector with a size of ~ 800kb with a total memory used by the wm_sys_main loop of approximately 1.5Mb.

https://github.com/wazuh/wazuh/blob/58b85c551096100e849f3d6ee1fa3876ed4a4010/src/data_provider/src/sysInfoWin.cpp#L547-L548

https://github.com/wazuh/wazuh/blob/58b85c551096100e849f3d6ee1fa3876ed4a4010/src/data_provider/src/packages/packagesWindowsParserHelper.h#L45-L46

https://github.com/wazuh/wazuh/blob/58b85c551096100e849f3d6ee1fa3876ed4a4010/src/data_provider/src/packages/packagesWindowsParserHelper.h#L80-L81

Regards,
Nico

### PR DESCRIPTION
**Pull request by konangc87**
|Related issue|
|---|
| Closes #9386  |

## Description

This PR aims to fix the issue found during memory analysis of syscollector related to the size of the returned std::vector by Registry::enumerate method.

## Tests

- Compilation without warnings in every supported platform
  - [x] ~Linux~
  - [ ] Windows
  - [x] ~MAC OS X~
- [x] ~Source installation~
- [ ] Package installation
- [x] ~Source upgrade~
- [ ] Package upgrade
- [ ] Review logs syntax and correct language
- [ ] QA templates contemplate the added capabilities

- Memory tests for Linux
  - [x] ~Scan-build report~
  - [x] ~Coverity~
  - [x] ~Valgrind (memcheck and descriptor leaks check)~
  - [x] ~Dr. Memory~
  - [x] ~AddressSanitizer~
- Memory tests for Windows
  - [ ] Scan-build report
  - [x] Coverity
  - [ ] Dr. Memory
- Memory tests for macOS
  - [x] ~Scan-build report~
  - [x] ~Leaks~
  - [x] ~AddressSanitizer~

- [ ] Retrocompatibility with older Wazuh versions
- [ ] Working on cluster environments
- [ ] Configuration on demand reports new parameters
- [ ] The data flow works as expected (agent-manager-api-app)
- [ ] ~Added unit tests (for new features)~
- [ ] Stress test for affected components

- Decoder/Rule tests
  - [X] ~Added unit testing files ".ini"~
  - [X] ~runtests.py executed without errors~

### CODE REVIEW

File: src/data_provider/src/sysInfoWin.cpp [hunk 1]
```diff
@@ -384,17 +384,20 @@ static void getPackagesFromReg(const HKEY key, const std::string& subKey, nlohma
...
-                packageJson["install_time"] = install_time;
-                packageJson["location"]     = location;
-                packageJson["architecture"] = architecture;
+                    packageJson["name"]         = std::move(name);
+                    packageJson["version"]      = std::move(version);
+                    packageJson["vendor"]       = std::move(vendor);
+                    packageJson["install_time"] = std::move(install_time);
+                    packageJson["location"]     = std::move(location);
+                    packageJson["architecture"] = std::move(architecture);
                 packageJson["format"]       = "win";
```
Comment (tinachoB) [inline src/data_provider/src/sysInfoWin.cpp:393]: bad indentation

**Review by tinachoB — APPROVED** [review]

**Comment by konangc87** [comment]
4.2 heap allocations in wazuh-agent.exe:
image
4.2 with hotfixes registry enum fix heap allocations in wazuh-agent.exe:
image

**Comment by jnasselle** [comment]
### Coverity execution for Windows Agent
<details>

```
6 new defect(s) introduced to wazuh/ossec-wazuh found with Coverity Scan.
29 defect(s), reported by Coverity Scan earlier, were marked fixed in the recent build analyzed by Coverity Scan.

New defect(s) Reported-by: Coverity Scan
Showing 6 of 6 defect(s)
** CID 221519:  Error handling issues  (CHECKED_RETURN)
/active-response/route-null.c: 136 in main()


________________________________________________________________________________________________________
*** CID 221519:  Error handling issues  (CHECKED_RETURN)
/active-response/route-null.c: 136 in main()
130                         os_free(gateway);
131                         os_strdup(ptr+1, gateway);
132                     }
133                 }
134                 fclose(fp);
135             }
>>>     CID 221519:  Error handling issues  (CHECKED_RETURN)
>>>     Calling "remove(tmp_file)" without checking return value. This library function may fail and return an error code.
136             remove(tmp_file);
137     
138             if (gateway) {
139                 snprintf(cmd, OS_MAXSTR, "%%WINDIR%%\\system32\\route.exe -p ADD %s MASK 255.255.255.255 %s", srcip, gateway);
140                 system(cmd);
141                 os_free(gateway);

** CID 221518:    (TAINTED_SCALAR)
/rootcheck/win-common.c: 106 in os_check_ads()
/rootcheck/win-common.c: 65 in os_check_ads()


________________________________________________________________________________________________________
*** CID 221518:    (TAINTED_SCALAR)
/rootcheck/win-common.c: 106 in os_check_ads()
100             /* Get next */
101             if (!BackupSeek(file_h, sid.Size.LowPart, sid.Size.HighPart,
102                             &dw1, &dw2, &context)) {
103                 break;
104             }
105         }
>>>     CID 221518:    (TAINTED_SCALAR)
>>>     Passing tainted expression "sid.dwStreamNameSize" to "BackupRead", which uses it as an offset.
106         BackupRead(file_h, (LPBYTE)stream_name,
107                            sid.dwStreamNameSize,
108                            &dwRead, TRUE, FALSE, &context);
109         CloseHandle(file_h);
110         return (0);
111     }
/rootcheck/win-common.c: 65 in os_check_ads()
59             if (dwRead == 0) {
60                 break;
61             }
62     
63             stream_name[0] = '\0';
64             stream_name[MAX_PATH] = '\0';
>>>     CID 221518:    (TAINTED_SCALAR)
>>>     Passing tainted expression "sid.dwStreamNameSize" to "BackupRead", which uses it as an offset.
65             if (BackupRead(file_h, (LPBYTE)stream_name,
66                            sid.dwStreamNameSize,
67                            &dwRead, FALSE, FALSE, &context)) {
68                 if (dwRead != 0) {
69                     DWORD i = 0;
70                     int max_path_size = 0;

** CID 221516:  Integer handling issues  (CONSTANT_EXPRESSION_RESULT)
/config/remote-config.c: 231 in Read_Remote()


________________________________________________________________________________________________________
*** CID 221516:  Integer handling issues  (CONSTANT_EXPRESSION_RESULT)
/config/remote-config.c: 231 in Read_Remote()
225                 const char * TIME_UNITS = "sSmMhHdD";
226     
227                 time_unit_ptr = strpbrk(node[i]->content, TIME_UNITS);
228                 rids_closing_time = w_parse_time(node[i]->content);
229     
230                 if ((time_unit_ptr != NULL && *(time_unit_ptr + 1) !='\0') ||
>>>     CID 221516:  Integer handling issues  (CONSTANT_EXPRESSION_RESULT)
>>>     "rids_closing_time > 2147483647" is always false regardless of the values of its operands. This occurs as the logical second operand of "||".
231                     (rids_closing_time <= 0 || rids_closing_time > INT_MAX)) {
232                         mwarn(REMOTED_INV_VALUE_DEFAULT, node[i]->content, xml_rids_closing_time);
233                         rids_closing_time = REMOTED_RIDS_CLOSING_TIME_DEFAULT;
234                 }
235     
236                 logr->rids_closing_time = (int) rids_closing_time;

** CID 221515:  Resource leaks  (RESOURCE_LEAK)
/win32/win_service.c: 197 in InstallService()


________________________________________________________________________________________________________
*** CID 221515:  Resource leaks  (RESOURCE_LEAK)
/win32/win_service.c: 197 in InstallService()
191                            MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
192                            (LPTSTR) &lpMsgBuf,
193                            0,
194                            NULL);
195     
196             merror("Unable to create service entry: %s", (LPCTSTR)lpMsgBuf);
>>>     CID 221515:  Resource leaks  (RESOURCE_LEAK)
>>>     Variable "lpMsgBuf" going out of scope leaks the storage it points to.
197             return (0);
198         }
199     }
200     
201     /* Uninstall the OSSEC-HIDS agent service */
202     int UninstallService()

** CID 214320:  Error handling issues  (CHECKED_RETURN)
/shared/read-agents.c: 847 in delete_diff()


________________________________________________________________________________________________________
*** CID 214320:  Error handling issues  (CHECKED_RETURN)
/shared/read-agents.c: 847 in delete_diff()
841         char tmp_folder[513];
842         tmp_folder[512] = '\0';
843         snprintf(tmp_folder, 512, "%s/%s",
844                  DIFF_DIR,
845                  name);
846     
>>>     CID 214320:  Error handling issues  (CHECKED_RETURN)
>>>     Calling "rmdir_ex" without checking return value (as is done elsewhere 5 out of 6 times).
847         rmdir_ex(tmp_folder);
848     }
849     
850     /* Delete agent */
851     int delete_agentinfo(const char *id, const char *name)
852     {

** CID 208487:  Insecure data handling  (TAINTED_SCALAR)


________________________________________________________________________________________________________
*** CID 208487:  Insecure data handling  (TAINTED_SCALAR)
/syscheckd/db/fim_db.c: 661 in fim_db_process_read_file()
655                 if (split == NULL || *split != ' ') {
656                     merror("Temporary path file '%s' is corrupt: Wrong format", file->path);
657                 } else {
658                     os_calloc(1, sizeof(fim_entry), entry);
659                     entry->type = FIM_TYPE_REGISTRY;
660     
>>>     CID 208487:  Insecure data handling  (TAINTED_SCALAR)
>>>     Passing tainted expression "arch" to "fim_db_get_registry_key", which uses it as an offset.
661                     entry->registry_entry.key = fim_db_get_registry_key(fim_sql, (split + 1), arch);
662                     if (entry->registry_entry.key == NULL) {
663                         free_entry(entry);
664                         entry = NULL;
665                     }
666                 }
```

</details>

### 6hs syscollector stress test

<details>

PR Job: [https://ci.wazuh.info/view/Tests/job/Test_stress/1887/](https://ci.wazuh.info/view/Tests/job/Test_stress/1887/)
4.2 Job: [https://ci.wazuh.info/view/Tests/job/Test_stress/1887/](https://ci.wazuh.info/view/Tests/job/Test_stress/1889/)

Despite stress test graphics does not show memory usage improvement, WPE shows lower heap allocation count and size.
 
- 4.2
monitor-winagent-Test_stress_B1889_windows-warehouse-branches_RSS

- Current PR
monitor-winagent-Test_stress_B1887_windows-warehouse-pullrequests_RSS

</details>

### Scanbuild

<details>


</details>

**Review by jnasselle — APPROVED** [review]

**Comment by Dwordcito** [comment]
No leaks detected, after 2nd scan (comparison between 1st and 2nd scan).

[raw_pr_fede-2.log](https://github.com/wazuh/wazuh/files/6895439/raw_pr_fede-2.log)


Application verifier no issues.
image

**Comment by Dwordcito** [comment]
Hotfix functional test: 

Windows 10 - 64bits:
image

Windows 10 - 32bits:
image

Windows 8 - 64bits:
image

Windows 8 - 32bits:
image

Windows 7 - 32bits:
image

Windows XP
image

**Review by Dwordcito — APPROVED** [review]
Looks good!

**Comment by Dwordcito** [comment]
@vikman90 ready for the merge

**Pull request merged**